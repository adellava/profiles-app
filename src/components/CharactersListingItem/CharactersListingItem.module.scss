@use 'src/styles' as style;

.CharactersListingItem {
  display: grid;
  padding: 1rem;
  grid-row-gap: 1rem;
  grid-column-gap: 1rem;
  grid-template-areas: 'img' 'description';
  grid-template-rows: auto auto;
  background-color: style.$light-primary;
  border-radius: style.$border-radius;
  @include style.elevation(2);
  &_action {
    display: inline;
    all: unset;
  }
  &_img {
    grid-area: img;
    aspect-ratio: 4/3;
    border-radius: style.$border-radius;
    object-fit: cover;
    width: 100%;
  }
  &_description {
    grid-area: description;
    background-color: style.$bg-color;
    color: style.$gray-800;
    border-radius: style.$border-radius;
    padding: 1rem;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    font-size: style.$font-size-600;
    font-weight: style.$text-standard-weight;
    text-align: left;
  }
  &_title {
    color: style.$black;
    font-size: style.$font-size-400;
    font-weight: style.$text-featured-weight;
  }
  &_meta {
    color: style.$gray-600;
    text-transform: uppercase;
    font-size: style.$font-size-800;
    font-weight: style.$text-featured-weight;
    margin-bottom: 0.3rem;
    margin-top: 0.7rem;
  }
  &_subtitle1 {
    margin-top: 0.2rem;
    color: style.$black;
    font-size: style.$font-size-500;
    font-weight: style.$text-standard-weight;
  }
  &_subtitle2 {
    margin-top: 0.2rem;
    color: style.$black;
    font-size: style.$font-size-500;
    font-weight: style.$text-standard-weight;
    font-style: italic;
  }
}

@include style.min-width(large) {
  .CharactersListingItem {
    grid-template-areas: 'img description';
    grid-template-columns: 1fr 1.5fr;
    grid-template-rows: 1fr;
    &_img {
      aspect-ratio: 1/1;
    }
  }
}
