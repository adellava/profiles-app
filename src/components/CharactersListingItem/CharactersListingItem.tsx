import { Character } from 'src/@types/Character';
import { GenderImage, StatusImage } from 'src/components';
import styles from './CharactersListingItem.module.scss';

type CharactersListingItemProps = {
  character: Character;
  onCharacterClick: (url: string) => void;
};

export const CharactersListingItem = ({
  character,
  onCharacterClick,
}: CharactersListingItemProps) => {
  return (
    <article
      key={character.id}
      className={styles.CharactersListingItem}
      onClick={() => onCharacterClick(character.url)}
    >
      <img
        src={character.image}
        alt={character.name}
        className={styles.CharactersListingItem_img}
      />
      <div className={styles.CharactersListingItem_description}>
        <section>
          <div>
            <h3 className={styles.CharactersListingItem_title}>
              <button className={styles.CharactersListingItem_buttonTitle}>
                {character.name}
              </button>
            </h3>
          </div>
          <div className={styles.CharactersListingItem_subtitle1}>
            {character.species} - <StatusImage statusType={character.status} />
            {character.status}
          </div>
          <div className={styles.CharactersListingItem_subtitle2}>
            <GenderImage genderType={character.gender} />
            {character.gender}
          </div>
        </section>

        <section>
          <div className={styles.CharactersListingItem_meta}>Location</div>
          <div>{character.location.name}</div>
        </section>

        <section>
          <div className={styles.CharactersListingItem_meta}>Origin</div>
          <div>{character.origin.name}</div>
        </section>

        <section>
          <div className={styles.CharactersListingItem_meta}>
            Number of Episodes
          </div>
          <div>{character.episode.length}</div>
        </section>
      </div>
    </article>
  );
};
