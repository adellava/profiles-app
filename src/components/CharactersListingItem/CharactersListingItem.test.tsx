import { CharactersListingItem } from './CharactersListingItem';
import TestRenderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
describe('CharactersListingItem', () => {
  it('should render', () => {
    expect(CharactersListingItem).toBeTruthy();
  });

  it('should render a character', () => {
    const testRenderer = TestRenderer.create(
      <CharactersListingItem
        character={{
          id: 1,
          gender: 'Male',
          name: 'Rick Sanchez',
          status: 'Alive',
          species: 'Human',
          type: '',
          episode: [],
          image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
          url: 'https://rickandmortyapi.com/api/character/1',
          origin: {
            name: 'Earth (C-137)',
            url: 'https://rickandmortyapi.com/api/location/1',
          },
          location: {
            name: 'Earth (Replacement Dimension)',
            url: 'https://rickandmortyapi.com/api/location/20',
          },
        }}
        onCharacterClick={() => {
          console.log('click');
        }}
      />
    );

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('should trigger onCharacterClick', async () => {
    const onCharacterClick = vi.fn();
    render(
      <CharactersListingItem
        character={{
          id: 1,
          gender: 'Male',
          name: 'Rick Sanchez',
          status: 'Alive',
          species: 'Human',
          type: '',
          episode: [],
          image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
          url: 'https://rickandmortyapi.com/api/character/1',
          origin: {
            name: 'Earth (C-137)',
            url: 'https://rickandmortyapi.com/api/location/1',
          },
          location: {
            name: 'Earth (Replacement Dimension)',
            url: 'https://rickandmortyapi.com/api/location/20',
          },
        }}
        onCharacterClick={onCharacterClick}
      />
    );

    await userEvent.click(screen.getByRole('article'));
    expect(onCharacterClick).toHaveBeenCalled();
  });
});
