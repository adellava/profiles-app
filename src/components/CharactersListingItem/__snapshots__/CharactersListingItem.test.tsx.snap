// Vitest Snapshot v1

exports[`CharactersListingItem > should render a character 1`] = `
<article
  className="_CharactersListingItem_73de65"
  onClick={[Function]}
>
  <img
    alt="Rick Sanchez"
    className="_CharactersListingItem_img_73de65"
    src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
  />
  <div
    className="_CharactersListingItem_description_73de65"
  >
    <section>
      <div>
        <h3
          className="_CharactersListingItem_title_73de65"
        >
          <button
            className="_CharactersListingItem_buttonTitle_73de65"
          >
            Rick Sanchez
          </button>
        </h3>
      </div>
      <div
        className="_CharactersListingItem_subtitle1_73de65"
      >
        Human
         - 
        <span
          aria-label="Alive"
          className="_StatusImage_5db57b"
          role="img"
        >
          ðŸŸ¢
        </span>
        Alive
      </div>
      <div
        className="_CharactersListingItem_subtitle2_73de65"
      >
        <span
          aria-label="Male"
          className="_GenderImage_d4cd8b"
          role="img"
        >
          â™‚
        </span>
        Male
      </div>
    </section>
    <section>
      <div
        className="_CharactersListingItem_meta_73de65"
      >
        Location
      </div>
      <div>
        Earth (Replacement Dimension)
      </div>
    </section>
    <section>
      <div
        className="_CharactersListingItem_meta_73de65"
      >
        Origin
      </div>
      <div>
        Earth (C-137)
      </div>
    </section>
    <section>
      <div
        className="_CharactersListingItem_meta_73de65"
      >
        Number of Episodes
      </div>
      <div>
        0
      </div>
    </section>
  </div>
</article>
`;
